README.rst
Internet Download Manager
---------------------------

Python Internet Download Manager Library


Installing
----------

Install and update using `pip3`_:

.. code-block:: text

    $ pip3 install internetdownloadmanager

Python 3 and newer.

.. _pip3: https://pip.pypa.io/en/stable/quickstart/


Example
----------------



.. code-block:: python

    from internetdownloadmanager import Downloader

    downloader = Downloader(worker=25,
                            part_size=1000000,
                            resumable=True)

    downloader.download(url="http://example.com/file",
                        path= 'not_required.extension')

Output:
-------

.. code-block:: python

    INFO:internetdownloadmanager.utils:
    File Path:filename.extension
    File Size: 377277402
    Worker: 25
    Calculated Part: 378

    WARNING:internetdownloadmanager.utils:Download Starting
    Status: %3.17 | Transfer Rate =  27274 Kb/s | Time Left: 108 sn | Time: 8 sn | AVG Speed=15169 Kb/s



* worker: Set how many threads work (default=25)
* part_size: Set piece size downloaded byte (default=1000000)
* resumable: Setting the download to be resumable (default=False)

If something goes wrong and the process is interrupted

.. code-block:: python

    downloader.resume('example.file_extention.resumable')

Support
-------

*   Python 3.x
*   Supports all operating systems

Links
-----

*   License: `Apache License <https://github.com/dinceraslancom/internetdownloadmanager/LICENSE.rst>`_
*   Code: https://github.com/dinceraslancom/internetdownloadmanager


HISTORY.rst
0.0.1 (2019-12-11)
++++++++++++++++++

* Birth!




internetdownloadmanager/utils.py
from math import ceil
import requests
from threading import Lock
import sys
from time import time
from json import loads, dumps
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

lock = Lock()


def warning(message):
    logger.warning(message)


def info(message):
    logger.info(message)


def error(message):
    logger.error(message)


def write_handler(file, path, part, offset, data, resumable):
    with lock:
        file_handler(file, path).seek(offset)
        file_handler(file, path).write(data)

        if resumable:
            with open(f'{path}.resumable', '+a') as log:
                log.write(f'/{part}')


def round_2(score):
    return round(score, 2)


def fix(score):
    return ceil(score)


def get_now():
    return time()


def set_part_total(file_size, part_size):
    return ceil(file_size / part_size)


def get_resume_info(path):
    with open(path, 'r') as file:
        data = file.read().split('*')

    resume_info = loads(data[0])
    resume_info['downloaded_parts'] = set(int(i) for i in data[1][1::].split('/'))
    resume_info['downloaded_parts_len'] = len(resume_info['downloaded_parts'])
    return resume_info


def get_ranged_data(url, start, end):
    return requests.get(url, headers={'range': f'bytes={start}-{end}'}, stream=True)


def get_file_info(url):
    response_head = requests.head(url)
    if response_head.status_code != 200:
        response_head.raise_for_status()
    _file_info = response_head.headers
    _type, _extension = _file_info['Content-Type'].split("/")

    if 'Content-Length' not in _file_info:
        raise Exception('Can not get Content-Length')

    _size = int(_file_info['Content-Length'])

    return dict(type=_type,
                size=_size,
                extension=_extension)


def url_handler(url):
    if not url:
        raise Exception('URL is Invalid')
    return url


def path_handler(url, extension):
    if url.endswith('/'):
        url = url[:-1]

    path = url.split('/')[-1]

    if len(path):
        path = path[:30]

    if 0 < len(path.split('.')[-1]) < 5:
        return path
    else:
        return f'{path}.{extension}'


def file_handler(file=None, path=None):
    if not file:
        try:
            file = open(path, 'r+b', buffering=0)
        except IOError:
            file = open(path, 'wb', buffering=0)

    return file


def timer(func):
    def decorator(*args, **kwargs):
        self = kwargs['slf']

        start_time = get_now()
        func(*args, **kwargs)
        end_time = get_now()

        take_seconds = end_time - start_time

        will_done = fix(take_seconds * (self._part_total - self._part_count) / self.worker)
        during = fix(get_now() - self._start_time)
        speed = fix((self.part_size / take_seconds) * self.worker / 125)  # (1000/8)

        self.downloaded_data_size += speed

        average_speed = fix(self.downloaded_data_size / self.avg_count)

        percent = round_2((self._part_count / self._part_total) * 100)
        # print()
        sys.stdout.write(
            f'\rStatus: %{percent} | Transfer Rate =  {speed} Kb/s | Time Left: {will_done} sn | Time: {during} sn | AVG Speed={average_speed} Kb/s ')
        sys.stdout.flush()

    return decorator


def write_file_info(self):
    with open(self._path+'.resumable', 'w') as log:
        log.write(dumps(dict(url=self._url,
                             path=self._path,
                             part_size=str(self.part_size),
                             worker=str(self.worker),
                             info=self.info,
                             resumable=self.resumable
                             )) + '*')


def print_file_info(self):
    info(f'\nFile Path:{self._path} \n'
         f'File Size: {self._file_info["size"]} \n'
         f'Worker: {self.worker} \n'
         f'Calculated Part: {self._part_total - len(self._downloaded)} \n'
         )


internetdownloadmanager/internetdownloadmanager.py
from concurrent.futures import ThreadPoolExecutor

from .utils import get_file_info, set_part_total, url_handler, path_handler, file_handler, \
    write_handler, timer, get_ranged_data, get_resume_info, write_file_info, print_file_info, get_now, error, warning


class Downloader:
    def __init__(self, worker=25, part_size=1000000, info=True, resumable=False, _downloaded=set(), _part_count=0):
        self.worker = worker
        self.part_size = part_size
        self.info = info
        self.resumable = resumable
        self._downloaded = _downloaded

        self._file = None
        self._url = None
        self._path = None
        self._file_info = None

        self.downloaded_data_size = 0
        self.avg_count = 0

        self._part_count = _part_count
        self._part_total = 0
        self._start_time = None

    @timer
    def _downloader(self, part, start, end, slf):
        self.avg_count += 1
        response = get_ranged_data(url=self._url, start=start, end=end)
        if response.status_code not in [206, 200]:
            error('Request not allowed\nClosing..')

        write_handler(file=self._file,
                      path=self._path,
                      part=part,
                      offset=start,
                      data=response.content,
                      resumable=self.resumable)

        self._part_count += 1

    def download(self, url=None, path=None):

        self._url = url_handler(url)

        self._file_info = get_file_info(self._url)

        self._path = path or path_handler(self._url, self._file_info['extension'])
        self._file = file_handler(path=self._path)

        self._part_total = set_part_total(self._file_info['size'], self.part_size)

        if self.info:
            print_file_info(self)

        if not self._downloaded and self.resumable:
            write_file_info(self)

        self._start_time = get_now()

        warning('Download Starting')
        with ThreadPoolExecutor(max_workers=self.worker) as executor:
            for part in (set(range(0, self._part_total)) - self._downloaded):
                executor.submit(self._downloader,
                                part,
                                self.part_size * part,
                                self.part_size * part + self.part_size,
                                slf=self)

        file_handler(self._file).close()

    @classmethod
    def resume(cls, path):
        resume_info = get_resume_info(path)

        _resume = cls(info=resume_info['info'],
                      worker=int(resume_info['worker']),
                      resumable=resume_info['resumable'],
                      part_size=int(resume_info['part_size']),
                      _downloaded=resume_info['downloaded_parts'],
                      _part_count=resume_info['downloaded_parts_len'])

        _resume.download(url=resume_info['url'], path=resume_info['path'])
